packages/cli/src/util/update-notifier/index.ts

```diff
@@ -0,0 +1,103 @@
+import semver from 'semver';
+import XDGAppPaths from 'xdg-app-paths';
+import { dirname, resolve as resolvePath } from 'path';
+// import { fileURLToPath } from 'url';
+import { outputJSON, readJSON } from 'fs-extra';
+import type { PackageJson } from '@vercel/build-utils';
+import { spawn } from 'child_process';
+
+interface UpdateNotifierConfig {
+  version: string;
+  expireAt: number;
+  notified: boolean;
+}
+
+/**
+ * Determines if it needs to check for a newer CLI version and returns the last
+ * detected version. The version could be stale, but still newer than the
+ * current version.
+ */
+export default async function updateNotifier({
+  cacheDir = XDGAppPaths('com.vercel.cli').cache(),
+  distTag = 'latest',
+  pkg,
+  updateCheckInterval = 1000 * 60 * 60 * 24 * 7, // 1 week
+  wait,
+}: {
+  cacheDir?: string;
+  distTag?: string;
+  pkg: PackageJson;
+  updateCheckInterval?: number;
+  wait?: boolean;
+}): Promise<string | undefined> {
+  const cacheFile = resolvePath(
+    cacheDir,
+    'update-notifier',
+    `${pkg.name}-${distTag}.json`
+  );
+
+  const loadCache = async (): Promise<UpdateNotifierConfig | undefined> => {
+    try {
+      return await readJSON(cacheFile);
+    } catch (e: any) {
+      // cache does not exist or malformed
+    }
+  };
+
+  let cache = await loadCache();
+
+  if (!cache || cache.expireAt < Date.now()) {
+    await new Promise<void>(resolve => {
+      // spawn the worker, wait for the worker to report it's ready, then
+      // signal the worker to fetch the latest version
+
+      const script = resolvePath(
+        dirname(__filename),
+        '..',
+        '..',
+        '..',
+        'dist',
+        'update-worker.js'
+      );
+      // const script = resolvePath(dirname(fileURLToPath(import.meta.url)), '..', '..', '..', 'dist', 'update-worker.js');
+
+      const worker = spawn(process.execPath, [script], {
+        stdio: ['pipe', 'pipe', 'pipe', 'ipc'],
+        windowsHide: true,
+      });
```

Comment:
**question (soft-blocking):** Why are we spawning a process for this?

Reply:
Its supposed to run in the background so it doesn't wait to check for an update before executing the user's command.

Reply:
That sounds reasonable, but currently the parent process is still waiting for the child process to exit. I can't figure out why, either. I tried removing event listeners and fiddling with spawn options, but it's still always waiting.

I'd be happy to pair on this. I want to figure out why the parent isn't letting the child be free.

It should also be noted that a bug in this process can cause orphaned node process to pile up on a machine. I managed to get several myself when debugging this.

Reply:
I figured it out and fixed the problem. Essentially the worker process would not disconnect the IPC channel until the check completed.

I believe the code originally worked, but when writing the tests, logic was changed and the dist version was not re-checked.

Since we currently do not have any tests that test the update notifier using the dist version, perhaps a few tests should be added?

Reply:
The parent is still waiting for the child, in my testing. It's really fast, but it does still happen.

I messed with this some more and I think I figured it out. I think it was a combination of:

- IPC was not disconnected until the check completed
- the `message` event handler returned a promise that waited for the check to complete

If I remove both issues, the parent no longer waits for the child at all. Here's an example:

```
process.on('message', msg => {
  process.disconnect();

  setImmediate(async () => {
    // pretend this is taking a while
    await new Promise((resolve) => {
      setTimeout(resolve, 10000);
    });

    const pending = fetchLatest(msg);
    if (msg.wait) {
      await pending;
    }
  });
});
```

Running this, I can see the initial command exit and an extra node process running for ~10 more seconds. After that process ends, I see the cache file populated.

Reply:
We discussed this on a call: the code as written does allow the parent to exit reasonably soon without waiting on the child. We decided to add a few more protections to ensure a clean system.